dnsext-full-resolver/dug/JSON.hs

```diff
@@ -0,0 +1,110 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module JSON (showJSON) where
+
+import DNS.Types
+import DNS.Types.Encode
+import qualified Data.ByteString.Base16 as B16
+import qualified Data.ByteString.Char8 as B
+import Data.Maybe
+import Prettyprinter
+import Prettyprinter.Render.String
+
+showJSON :: DNSMessage -> String
+showJSON DNSMessage{..} = toString $ object (hd ++ q ++ [an, au, ad])
+  where
+    DNSHeader{..} = header
+    DNSFlags{..} = flags
+    hd =
+        [ ("ID", pretty identifier)
+        , ("QR", pretty $ fromQR qOrR)
+        , ("Opcode", pretty $ fromOPCODE opcode)
+        , ("comment", string opcode)
+        , ("AA", pretty $ fromBool authAnswer)
+        , ("TC", pretty $ fromBool trunCation)
+        , ("RD", pretty $ fromBool recDesired)
+        , ("RA", pretty $ fromBool recAvailable)
+        , ("AD", pretty $ fromBool authenData)
+        , ("CD", pretty $ fromBool chkDisable)
+        , ("RCode", pretty $ fromRCODE rcode)
+        , ("comment", string rcode)
+        , ("QDCOUNT", pretty $ length question)
+        , ("ANCOUNT", pretty $ length answer)
+        , ("NSCOUNT", pretty $ length authority)
+        , ("ARCOUNT", pretty $ length additional)
+        ]
+    q = case question of
+        [] -> []
+        (Question{..} : _) ->
+            [ ("QNAME", domain qname)
+            , ("QTYPE", pretty $ fromTYPE qtype)
+            , ("QTYPEname", string qtype)
+            , ("QCLASS", pretty qclass)
+            ]
+    an = rrs "answerRRs" answer
+    au = rrs "authorityRRs" authority
+    ad = rrs "additionalRRs" additional
+
+toString :: Doc ann -> String
+toString = renderString . layoutPretty defaultLayoutOptions
+
+object :: [(String, Doc ann)] -> Doc ann
+object [] = align (lbrace <> rbrace)
+object kvs = align (vsep ([lbrace] ++ punctuate comma pairs ++ [rbrace]))
+  where
+    pairs = map pair kvs
+
+array :: [Doc ann] -> Doc ann
+array [] = align (lbracket <> rbracket)
+array objs = align (vsep ([lbracket] ++ punctuate comma objs ++ [rbracket]))
+
+rrs :: String -> [ResourceRecord] -> (String, Doc ann)
+rrs k xs = (k, array $ map rr xs)
+
+rr :: ResourceRecord -> Doc ann
+rr ResourceRecord{..} = object xs
+  where
+    Seconds ttl = rrttl
+    xs =
+        [ ("NAME", domain rrname)
+        , ("TYPE", pretty $ fromTYPE rrtype)
+        , ("TYPEname", string rrtype)
+        , ("TTL", pretty ttl)
+        , rd rrtype rdata
+        ]
+
+{- FOURMOLU_DISABLE -}
+rd :: TYPE -> RData -> (String, Doc ann)
+rd A     v = ("rdA",         justQuote (fromRData v :: Maybe RD_A))
+rd AAAA  v = ("rdAAAA",      justQuote (fromRData v :: Maybe RD_AAAA))
```

Comment:
```suggestion
rd A     v = ("rdataA",      justQuote (fromRData v :: Maybe RD_A))
rd AAAA  v = ("rdataAAAA",   justQuote (fromRData v :: Maybe RD_AAAA))
```

adjust field members to https://datatracker.ietf.org/doc/html/rfc8427#section-2.3 ?

Reply:
[5aab4f4](https://github.com/kazu-yamamoto/dnsext/pull/115/commits/5aab4f459ade2282a521e585a704b429413d503f) fixes them